h/static/scripts/controllers/autosuggest-dropdown-controller.js

```diff
@@ -0,0 +1,454 @@
+   */
+  _keyboardSelectionChange(down){
+
+    const currentActive = this._getActiveListItemElement();
+    let nextActive;
+
+    // we have a starting point, navigate on siblings of current
+    if(currentActive){
+      currentActive.classList.remove(this.options.classNames.activeItem);
+
+      if(down){
+        nextActive = currentActive.nextSibling;
+      }else {
+        nextActive = currentActive.previousSibling;
+      }
+
+    // we have no starting point, let's navigate based on
+    // the directional expectation of what the first item would be
+    }else if(down){
+      nextActive = this._listContainer.firstChild;
+    }else {
+      nextActive = this._listContainer.lastChild;
+    }
+
+    if(nextActive){
+      nextActive.classList.add(this.options.classNames.activeItem);
+    }
+  }
+
+  /**
+   * _renderContentContainers - build the DOM structure that makes up
+   *  the suggestion box and content containers.
+   */
+  _renderContentContainers(){
+
+    // container of all suggestion elements
+    this._suggestionContainer = document.createElement('div');
+    this._suggestionContainer.classList.add(this.options.classNames.container);
+
+
+    // child elements that will be populated by consumer
+    this._header = document.createElement('h4');
+    this._header.classList.add(this.options.classNames.header);
+    this._setHeader(this.state.header);
+    this._suggestionContainer.appendChild(this._header);
+
+    this._listContainer = document.createElement('ul');
+    this._listContainer.classList.add(this.options.classNames.list);
+    this._suggestionContainer.appendChild(this._listContainer);
+
+    // put the suggestions adjacent to the input element
+    // firefox does not support insertAdjacentElement
+    if(HTMLElement.prototype.insertAdjacentElement){
+      this._input.insertAdjacentElement('afterend', this._suggestionContainer);
+    }else {
+      this._input.parentNode.insertBefore(this._suggestionContainer, this._input.nextSibling);
+    }
+  }
+
+  /**
+   * _renderListItems - updates the content of the list container and builds
+   *  the new set of list items.
+   */
+  _renderListItems(){
+    // Create the new list items, render their contents
+    // and update the dom with the new elements.
+
+    const currentActive = this._getActiveListItemElement();
+    const currentActiveSelectorId = currentActive ? currentActive.getAttribute('data-suggestion-id') : null;
+
+    this._listContainer.innerHTML = '';
+
+    this.state.list.forEach((listItem)=>{
+      let li = document.createElement('li');
+      li.classList.add(this.options.classNames.item);
+      li.setAttribute('data-suggestion-id', listItem.__suggestionId);
+
+      // persist the active state of selections
+      // through the rendering process
+      if(listItem.__suggestionId === currentActiveSelectorId){
+        li.classList.add(this.options.classNames.activeItem);
+      }
+
+      // this should use some sort of event delegation if
```

Comment:
Controller's have a convenience `this.on(eventName, callback)` you can use if you wish. I agree with the comment here, given how few elements we're going to display, event delegation is not worth worrying about.